[SimplifyCFG] Transform switch to select when common bits uniquely identify one case

spaits opened the pull request:
My attempt to fix #141753 .

This patch introduces a new check, that tries to decide if the conjunction of all the values uniquely identify the accepted values by the switch.

I am also open to different solution. If you think this approach is not the best and have a different idea I would also be happy to implement that.

I was also thinking about addressing this pattern in inst combine. There was a previous commit, that adresses a very similar pattern to this https://github.com/llvm/llvm-project/commit/902acde34198bb11cc758dcf3aee00eb1cb09ceb, 
but I think it would be better if we wouldn't even get to the point of emitting this pattern.

spaits committed:
Transform switch to select when common bits uniquely identify one case

antoniofrighetto commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -6290,10 +6290,37 @@ static Value *foldSwitchToSelect(const SwitchCaseResultVectorTy &ResultVector,
+        AndMask &= Case->getValue();
+      }
+
+      ConstantRange CR = computeConstantRange(
+          Condition, /* ForSigned */ Condition->getType()->isSingleValueType());
```
What's exactly the reasoning of using `Condition->getType()->isSingleValueType()` for signedness flag? Not immediately clear to me. And wouldn't computeKnownBits work here?

spaits commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -6290,10 +6290,37 @@ static Value *foldSwitchToSelect(const SwitchCaseResultVectorTy &ResultVector,
+                               Condition->getType()->isSingleValueType());
+
+      APInt One(ConditionWidth, 1, false);
+      APInt ActiveBitsMask = (One << ActiveBits) - 1;
```
Wait, think about it more, we don't need it.
The bits coming from the argument are in a certain range. The bits set outside that range couldn't cause any problem, since we are `and` ing the argument to the mask.

spaits commented:
Thank you @antoniofrighetto @dtcxzyw . I think I have addressed everything. I have replaced `computeConstantRange` with `computeKnownBits` and removed the unused `ActiveBitsMask`.

spaits commented:
Sqashed the commits and re-run the CI (Some lldb tests have randomly failed due to possible env issues).

dtcxzyw commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -6234,10 +6234,34 @@ static Value *foldSwitchToSelect(const SwitchCaseResultVectorTy &ResultVector,
+      KnownBits Known = computeKnownBits(Condition, DL);
+      unsigned int ConditionWidth = Condition->getType()->getIntegerBitWidth();
+      APInt ActiveBits = APInt(ConditionWidth, Known.countMaxActiveBits(),
+                               Condition->getType()->isSingleValueType());
```
```suggestion
                               Condition->getType()->isSingleValueType());
```
It is evaluated as `true`. But it should be treated as an unsigned value.

dtcxzyw commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -6234,10 +6234,34 @@ static Value *foldSwitchToSelect(const SwitchCaseResultVectorTy &ResultVector,
+      // actually unique we check, wheter the conjucted bits and the another
+      // conjuction with the input value will only be true for exactly CaseCount
+      // number times.
+      if ((One << ActiveBits) - (One << (ActiveBits - AndMask.popcount())) ==
```
Looks like `if (isPowerOf2_32(CaseCount))` is unnecessary for this transform. `isPowerOf2(CaseCount) && (1 << ActiveBits) - (1 << (ActiveBits - AndMask.popcount())) == CaseCount` implies `AndMask.popcount() == 1`.

spaits replied:
I have implemented the `cmp eq and(%1, AndMask) AndMask` version. Also chenged the guard condition accordingly. Also extended testing.

spaits committed:
ActiveBits shall be unsigned regardless of the Condition

spaits committed:
Generate identifier pattern using  and add more tests

spaits commented:
@dtcxzyw Thank you very much for reviewing my PR.
- I have rebased the PR to main
- ActiveBits is now unsigned
- Replaced the old `icmp ne (and %1 AndMask) 0` with `cmp eq and(%1, AndMask) AndMask`. You have pointed out, that the old method could only identify `select`s where there was one single bit common in every pattern. The new method can use multiple bits to identify the pattern. A good example of such a case, where the new method works but the old doesn't is test `range0to15_middle_two_bits`. I have also added some negative tests too.

Thank you very much for reviewing and pointing out the inefficiency in my original logic.
I would like to ask you to please re-review the PR.

antoniofrighetto commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -6234,10 +6234,33 @@ static Value *foldSwitchToSelect(const SwitchCaseResultVectorTy &ResultVector,
+      }
+
+      KnownBits Known = computeKnownBits(Condition, DL);
+      unsigned int ConditionWidth = Condition->getType()->getIntegerBitWidth();
```
```suggestion
      unsigned ConditionWidth = Condition->getType()->getIntegerBitWidth();
```

antoniofrighetto commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -6234,10 +6234,33 @@ static Value *foldSwitchToSelect(const SwitchCaseResultVectorTy &ResultVector,
+
+      KnownBits Known = computeKnownBits(Condition, DL);
+      unsigned int ConditionWidth = Condition->getType()->getIntegerBitWidth();
+      APInt ActiveBits =
+          APInt(ConditionWidth, Known.countMaxActiveBits(), false);
+
+      APInt One(ConditionWidth, 1, false);
+      // To make sure, that the representation of the accepted values is
+      // actually unique we check, wheter the conjucted bits and the another
+      // conjuction with the input value will only be true for exactly CaseCount
+      // number times.
+      if ((One << (ActiveBits - AndMask.popcount())) == CaseCount) {
```
I feel like there are too unnecessary APInts here, what about the following? I think this should be safe from wrap-arounds too (should never have the number of bit sets in AndMask greater than the active bits).
```cpp
KnownBits Known = computeKnownBits(Condition, DL);
unsigned ConditionWidth = Condition->getType()->getIntegerBitWidth();

// To make sure that the representation of the accepted values is
// actually unique, we check whether the conjuction bits and the other
// conjuction with the input value will only be true for exactly CaseCount
// number times.

// Compute the number of bits that are free to vary.
unsigned FreeBits = Known.countMaxActiveBits() - AndMask.popcount();
// Compute 2^FreeBits in order to check whether all the possible
// combination of the free bits matches the number of cases.
APInt TopBit = APInt::getOneBitSet(ConditionWidth, FreeBits);

if (TopBit == CaseCount) {
   // ...
}
```

spaits replied:
This is a really elagant way to do the check! Much nicer than my current one, which computes powers of two with shifting. I will swicth to it. Thank you for recommending this.

spaits replied:
`FreeBits` could onyl be bigger than `ConditionWidth`, if the `Known.countMaxActiveBits() - AndMask.popcount()` overflows, so we could just cut of all the possible 1s, that are outside of active bits.

I think this would be a legal move. It doesn't matter what is in `AndMask` before the active bits. That part of the parameters would be 0 anyway.

What do you think?

antoniofrighetto replied:
I don't think `Known.countMaxActiveBits() - AndMask.popcount()` can ever wrap around. FreeBits may be larger than ConditionWidth when `AndMask` is zero. Should probably check `!AndMask.isZero()` at the beginning to proceed with the transform.

spaits committed:
unsigned int -> unsigned

spaits committed:
Use new way to compute if pattern identifies all

spaits committed:
Add out of range tests

antoniofrighetto commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -6234,10 +6234,35 @@ static Value *foldSwitchToSelect(const SwitchCaseResultVectorTy &ResultVector,
       ConstantInt *MinCaseVal = CaseValues[0];
-      // Find mininal value.
-      for (auto *Case : CaseValues)
+      // In case, there are bits, that can only be present in the CaseValues we
+      // can transform the switch into a select if the conjunction of
+      // all the values uniquely identify the CaseValues.
```
```suggestion
      // If there are bits that are set exclusively by CaseValues, we
      // can transform the switch into a select if the conjunction of
      // all the values uniquely identify CaseValues.
```

antoniofrighetto commented on llvm/test/Transforms/SimplifyCFG/switch-to-select-two-case.ll:
```diff
@@ -309,3 +309,291 @@ end:
+bb2:
+  br label %bb3
+bb3:
+  %_0.sroa.0.0 = phi i1 [ false, %bb1 ], [ true, %bb2 ]
+  ret i1 %_0.sroa.0.0
```
```suggestion
  %phi = phi i1 [ false, %bb1 ], [ true, %bb2 ]
  ret i1 %phi
```
Same below for the others.

antoniofrighetto commented on llvm/test/Transforms/SimplifyCFG/switch-to-select-two-case.ll:
```diff
@@ -309,3 +309,291 @@ end:
+  %_0.sroa.0.0 = phi i1 [ false, %bb1 ], [ true, %bb2 ]
+  ret i1 %_0.sroa.0.0
+}
+
```
Nit: drop extra newline.

antoniofrighetto commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -6234,10 +6234,35 @@ static Value *foldSwitchToSelect(const SwitchCaseResultVectorTy &ResultVector,
+        // Compute the number of bits that are free to vary.
+        unsigned FreeBits = Known.countMaxActiveBits() - AndMask.popcount();
+        // Compute 2^FreeBits in order to check whether all the possible
+        // combination of the free bits matches the number of cases.
```
```suggestion
        // combinations of the free bits matches the number of cases.
```

antoniofrighetto reviewed (approved):
LGTM, please wait for additional approval.

spaits committed:
All out of range case

spaits committed:
Extend testing

spaits committed:
Address review findings

dtcxzyw commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -6234,10 +6234,35 @@ static Value *foldSwitchToSelect(const SwitchCaseResultVectorTy &ResultVector,
+        APInt TopBit = APInt::getOneBitSet(
+            Condition->getType()->getIntegerBitWidth(), FreeBits);
+
+        if (TopBit == CaseCount) {
```
```suggestion
        if (FreeBits == Log2_32(CaseCount)) {
```

dtcxzyw reviewed (approved):
LGTM. We may generalize this pattern in the future since some "free" bits may be known. But I think the current implementation is enough to cover the motivating cases.

spaits committed:
Use Log3_32
